tool/tctl/common/device/resource.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package device
+
+import (
+	"github.com/gravitational/trace"
+
+	devicepb "github.com/gravitational/teleport/api/gen/proto/go/teleport/devicetrust/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// Resource is a wrapper around devicepb.Device that implements types.Resource.
+type Resource struct {
+	// ResourceHeader is embedded to implement types.Resource
+	types.ResourceHeader
+	// Spec is the device specification
+	Spec *devicepb.Device `json:"spec"`
+}
+
+// CheckAndSetDefaults sanity checks Resource fields to catch simple errors, and
+// sets default values for all fields with defaults.
+func (r *Resource) CheckAndSetDefaults() error {
+	if err := r.Metadata.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+	if r.Kind == "" {
+		r.Kind = types.KindDevice
+	} else if r.Kind != types.KindDevice {
+		return trace.BadParameter("unexpected resource kind %q, must be %q", r.Kind, types.KindDevice)
+	}
+	if r.Spec == nil {
+		return trace.BadParameter("device must have a spec")
+	}
+	switch {
+	case r.Version != types.V1:
+		return trace.BadParameter("unsupported resource version %q, %q is currently the only supported version", r.Version, types.V1)
+	case r.Spec.ApiVersion == "":
+		r.Spec.ApiVersion = r.Version
+	case r.Spec.ApiVersion != types.V1:
+		return trace.BadParameter("mismatched resource version %q and spec api version %q", r.Version, r.Spec.ApiVersion)
+	}
+	switch {
+	case r.Metadata.Name == "":
```

Comment:
I took a look at this. From my understanding it's possible to have two devices with the same asset tag with this specification? If so, that'd cause problems since `Metadata.Name` shall be unique. Is this the case?

Reply:
I reworked the resource a bit and IMO it's quite a bit cleaner to store the ID in the metadata name as it needs to be unique and keep the asset tag as part of the spec. WDYT?